internal/reminder/reminder.go

```diff
@@ -0,0 +1,197 @@
+type Interface interface {
+	// Start starts the reminder by sending reminders at regular intervals
+	Start(ctx context.Context) error
+
+	// Stop stops the reminder service
+	Stop()
+}
+
+// reminder sends reminders to the minder server to process entities in background.
+type reminder struct {
+	store    db.Store
+	cfg      *reminderconfig.Config
+	stop     chan struct{}
+	stopOnce sync.Once
+
+	projectListCursor string
+	repoListCursor    map[projectProviderPair]string
+
+	ticker *time.Ticker
+	logger zerolog.Logger
+
+	eventPublisher message.Publisher
+	eventDBCloser  driverCloser
+}
+
+type projectProviderPair struct {
+	// Exported for gob
+
+	ProjectId uuid.UUID
+	Provider  string
+}
+
+// NewReminder creates a new reminder instance
+func NewReminder(store db.Store, config *reminderconfig.Config) (Interface, error) {
+	level := util.ViperLogLevelToZerologLevel(config.LoggingLevel)
+	r := &reminder{
+		store:          store,
+		cfg:            config,
+		stop:           make(chan struct{}),
+		repoListCursor: make(map[projectProviderPair]string),
+		logger:         zerolog.New(os.Stdout).Level(level).With().Timestamp().Logger(),
+	}
+	err := r.restoreCursorState()
+	if err != nil {
+		// Non-fatal error, if we can't restore the cursor state, we'll start from scratch.
+		r.logger.Error().Err(err).Msg("error restoring cursor state")
+	}
+
+	pub, cl, err := r.setupSQLPublisher(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	r.eventPublisher = pub
+	r.eventDBCloser = cl
+	return r, nil
+}
+
+// Start starts the reminder by sending reminders at regular intervals
+func (r *reminder) Start(ctx context.Context) error {
+	select {
+	case <-r.stop:
+		return errors.New("reminder stopped, cannot start again")
+	default:
+	}
+
+	interval, err := time.ParseDuration(r.cfg.RecurrenceConfig.Interval)
+	if err != nil {
+		return err
+	}
+
+	if interval <= 0 {
+		return fmt.Errorf("invalid interval: %s", r.cfg.RecurrenceConfig.Interval)
+	}
+
+	r.ticker = time.NewTicker(interval)
+	defer r.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.stop:
+			r.logger.Info().Msg("reminder stopped")
+			return nil
+		case <-r.ticker.C:
+			// In-case sending reminders i.e. iterating over entities consumes more time than the
+			// interval, the ticker will adjust the time interval or drop ticks to make up for
+			// slow receivers.
+			if errs := r.sendReminders(ctx); errs != nil {
+				for _, err := range errs {
+					r.logger.Error().Err(err).Msg("reconciliation request unsuccessful")
+				}
+			}
+		}
+	}
+}
+
+// Stop stops the reminder service
+func (r *reminder) Stop() {
+	if r.ticker != nil {
+		defer r.ticker.Stop()
+	}
+	r.stopOnce.Do(func() {
+		close(r.stop)
+		r.eventDBCloser()
+	})
+}
+
+func (r *reminder) storeCursorState() error {
+	r.logger.Debug().Msg("storing cursor state")
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
```

Comment:
Why use `gob` rather than simply JSON?

Reply:
How large is this data going to be?  I see that it's stored in a `map`, so I'm wondering how many keys there will be in the map.

Reply:
It is directly proportional to the number of projects being processed in an iteration, which is configurable, so it can get as large as you'd want

Reply:
I based it on: https://github.com/RSheremeta/gob-serialization/blob/master/Analysis.md

Reply:
Why do we need to keep track of all the projects processed in an iteration?  It should be fine to re-run a small number of projects, so we could keep just the first (or last) item in the batch.

Binary formats are a pain for humans to work with (we need to build extra tools), so I'd prefer JSON or YAML if possible so we can clean up intermediate state in an emergency with simple tools.

Reply:
> Why do we need to keep track of all the projects processed in an iteration? It should be fine to re-run a small number of projects, so we could keep just the first (or last) item in the batch.

We have project cursor and repo cursor for every project we traverse. The repo cursor list grows as we traverse more projects because we only pick small no of repos from a project to have sort of fair share type algo and avoid hitting rate limits for that project. 

We have to keep track of all repo cursors as when we return to the same project in subsequent loops, we want to pick the repos that haven't been picked yet.

> Binary formats are a pain for humans to work with (we need to build extra tools), so I'd prefer JSON or YAML if possible so we can clean up intermediate state in an emergency with simple tools.

I just had some performance concerns (though they may not matter that much), I'll change to json or yaml.

Reply:
> > Why do we need to keep track of all the projects processed in an iteration? It should be fine to re-run a small number of projects, so we could keep just the first (or last) item in the batch.
> 
> We have project cursor and repo cursor for every project we traverse. The repo cursor list grows as we traverse more projects because we only pick small no of repos from a project to have sort of fair share type algo and avoid hitting rate limits for that project.

If we were to traverse the lists in order of repo ID and rely on _Minder_ to limit reconciliations during contention, would that serve the same result?  I'm worried that we only have a partial view of what's going on in terms of quota usage here, so I'd rather have Reminder act as a reliable clock but not get too fancy in terms of trying to push back because Minder will still need to deal with quota exhaustion in other cases.  (We had one about a week ago due to poor cache behavior, for example.)